### ISSUE TITLE
Native M1 support

### ISSUE THREAD

**Issue by srid** [purs-nix/purs-nix#17]
It would be good to support M1 natively without having to use rosetta.

The main blocker is that PureScript releases ([used here](https://github.com/purs-nix/purescript-tools/tree/master/purescript)) don't include a M1 build. 

`purs-nix` does work under Rosetta, though (`--option system x86_64-darwin`), so that's a reasonable workaround in the interim.

**Comment by srid** [comment]
https://github.com/NixOS/nixpkgs/pull/198827

**Comment by toastal** [comment]
More recent versions (see: https://github.com/purs-nix/purescript-tools/blob/master/purescript/0.15.4.nix) cover this and throw a warning that it's not a fully-supported platform because it is running under a translation layer rather than being built for the platform (same reasoning listed in the Nixpkgs bug). If anything, I'd love to see the PureScript project more generally do ARM support rather than than just Darwin--or make the project Nixpkgs-compatible again so Hydra can build it.

**Comment by srid** [comment]
@toastal Note: `pkgs.purescript` (what you linked to above) is a different package to `pkgs.haskellPackages.purescript`. The latter is what is known to build successfully.

**Comment by toastal** [comment]
You sure? The maintainer says it's patched from upstream (https://github.com/NixOS/nixpkgs/pull/198827#pullrequestreview-1162490462). It also ignores the history of why that happened (https://github.com/justinwoo/easy-purescript-nix/blob/master/README.md#why-was-this-made). If it builds, why is it in the `broken.yaml` for Haskell (https://github.com/NixOS/nixpkgs/blob/38d81f21326497c7f1833b7e7f395aa8f8c6ab17/pkgs/development/haskell-modules/configuration-hackage2nix/broken.yaml#L4121). I'm pretty sure it's in a perpetual state of breaking and the PureScript team isn't interested in trying to keep up with Nix. Maybe someone found a way to make it work?

**Comment by srid** [comment]
Yes, what I meant by "different package" (bad phrasing, btw) was that they are both the same package, PureScript, but one is a binary downloaded off GitHub and other is built from source using nixpkgs's Haskell infrastructure.

The latter was marked as broken [only as of last month](https://github.com/NixOS/nixpkgs/commit/a53a14bc2df5e9975fd8352cddb5b343d62daa72) but it has been building before. 

From your comment elsewhere:
> Unsure about `haskellPackages.purescript`'s support level since I _believe_ the maintainers prefer the binaries that are downloaded and patched from PureScript's releases.

Could you add some references for this belief? What are patches are applied on released binaries (and why are not part of the source tarball)?

**Comment by srid** [comment]
Note: it is marked as broken only because[^err] the default packaget set (`pkgs.haskellPackages` set) is still using GHC 9.0.2, which is no longer being supported by PureScript (https://github.com/purescript/purescript/pull/4351).

`pkgs.haskell.packages.ghc925.purescript` does build, at least according to https://nixpkgs.haskell.page/9.2.5/p/purescript

[^err]:

    Run this on ARM mac:
    
    ```
    ❯ NIXPKGS_ALLOW_BROKEN=1 nix run --impure nixpkgs#haskellPackages.purescript -- --version
    error: builder for '/nix/store/y79wfy9kid7i050n494n6dzd7i9a7ykf-purescript-0.15.6.drv' failed with exit code 1;
           last 10 log lines:
           > Dependency transformers-compat: using transformers-compat-0.6.6
           > Dependency typed-process: using typed-process-0.2.10.1
           > Dependency unordered-containers: using unordered-containers-0.2.17.0
           > Dependency utf8-string: using utf8-string-1.0.2
           > Dependency vector: using vector-0.12.3.1
           > Dependency witherable: using witherable-0.4.2
           >
           > Setup: The package purescript-0.15.6 requires the following language
           > extensions which are not supported by ghc-9.0.2: OverloadedRecordDot
           >
           For full logs, run 'nix log /nix/store/y79wfy9kid7i050n494n6dzd7i9a7ykf-purescript-0.15.6.drv'.
    ```

**Comment by oati** [comment]
`--option system x86_64-linux` isn't enough to get purs-nix working on NixOS on M1. This is with
```nix
boot.binfmt.emulatedSystems = [ "x86_64-linux" ];
```
set in my configs.

**Comment by oati** [comment]
What would it take to be able to use, for example, `pkgs.haskell.packages.ghc925.purescript` with purs-nix on `aarch64-linux`? What would I need to change?

**Comment by srid** [comment]
AFAIU:

See https://github.com/purs-nix/purescript-tools/tree/master/purescript

For eg., https://github.com/purs-nix/purescript-tools/blob/master/purescript/0.15.4.nix could probably just be `{ pkgs }: pkgs.haskell.packages.ghc925.purescript` .. but since there is exactly one PureScript compiler version in the given nixpkgs, it will be tricky to support multiple versions this way .. unless perhaps purs-nix switches to a single compiler version mode (tangential: #31).

That, along with adding the new system to all relevant places, starting with https://github.com/purs-nix/purs-nix/search?q=x86_64-darwin

**Comment by oati** [comment]
> @oati Regarding your previous comment, I'm not sure why you have to emulate Linux itself.
> 
> Try https://github.com/Platonic-Systems/purescript-multi-nix which does work well with `--option system x86_64-darwin` on my M1 mac.

I'm running NixOS natively on my M1 mac. I'm not using Nix on macOS.

**Comment by ursi** [comment]
https://discourse.purescript.org/t/new-arm-builds-for-the-compiler/3500

**Comment by malikwirin** [comment]
purescript-tools should use upstream nixpkgs purescript as this one does build on `aarch64-darwin`

**Comment by ursi** [comment]
this should work now
